tests/test_resolver.py

```diff
@@ -0,0 +1,129 @@
+"""Verify behavior of packages resolver"""
+
+from typing import Dict, List, Optional
+
+from hypothesis import HealthCheck, assume, given, settings
+from hypothesis import strategies as st
+
+from fawltydeps.packages import DependenciesMapping, Package, resolve_dependencies
+
+
+def dict_subset_strategy(input_dict):
+    keys_strategy = st.lists(st.sampled_from(list(input_dict.keys())), unique=True)
+
+    def build_dict(keys):
+        return {key: input_dict[key] for key in keys}
+
+    return keys_strategy.map(build_dict)
+
+
+# The deps in each category should be disjoint
+
+# non locally installed deps
+non_locally_installed_deps = ["pandas", "numpy", "other"]
+non_locally_installed_strategy = st.lists(
+    st.sampled_from(non_locally_installed_deps), unique=True
+)
+
+# locally installed deps
+locally_installed_deps = {
+    "setuptools": [
+        "_distutils_hack",
+        "pkg_resources",
+        "setuptools",
+    ],
+    "pip": ["pip"],
+    "isort": ["isort"],
+}
+locally_installed_strategy = dict_subset_strategy(locally_installed_deps)
+
+# user-defined deps
+user_defined_deps = {"apache-airflow": ["airflow"]}
+user_defined_strategy = dict_subset_strategy(user_defined_deps)
+
+# user-defined mapping
+user_defined_mapping = "\n".join(
+    [f'{dep} = ["{",".join(imports)}"]' for dep, imports in user_defined_deps.items()]
+)
+
+user_mappings = st.one_of(st.none(), st.just(user_defined_mapping))
+
+
+def generate_expected_resolved_deps(
+    locally_installed_deps: Optional[Dict[str, List[str]]] = None,
+    non_locally_installed_deps: Optional[List[str]] = None,
+    user_defined_deps: Optional[Dict[str, List[str]]] = None,
+):
+    ret = {}
```

Comment:
From current implementation I get following resolution importance order:
- Identity Mapping
- User-defined mapping
- local env



In `resolve_dependencies` function we have:
- User-defined mapping
- local env
- temporary pip-install resolver
- Identity Mapping

What do you think about it?
It would be also helpful here to add a docstring that says what is the resolution order. I know it repeats what is in `resolved_deps`, but `generate_expected_resolved_deps` should just emulate the real function's behaviour.

Reply:
So the `generate_expected_resolved_deps` function does not actually resolve dependencies. It just replaces the generation of the "expected" dict that we were manually specifying in the parameterized tests.
There's no order for how the different categories are resolved. The resulting dictionary is just constructed from the different categories of deps.

To make this clearer, I added a docstring to `generate_expected_resolved_deps` in [e70a35a](https://github.com/tweag/FawltyDeps/pull/296/commits/e70a35ac22f13cdadbb92a46f9161f6106c34330) and added asserts to the test to ensure the condition of no overlap between different categories is respected in [e7443ab](https://github.com/tweag/FawltyDeps/pull/296/commits/e7443ab7a4ec6d04370d903794a92ffecd09db39).

Reply:
Thank you, now and after the pair programming discussion I know we were thinking about validation checks in different places. You made sure that the dependencies sets were disjoint in the main test function, and I wanted to make sure that we have a good resolution order not knowing that `generate_expected_resolver_deps` already tasks a disjoint sets.

With the current version, all is good :+1: